[WFLY-12298] Implement RESTEasy context parameters in Wildfly

ronsigal opened the pull request:
management model.

https://issues.jboss.org/browse/WFLY-12298 "Allow to change RESTEasy settings via CLI"

[WFLY-12298] Add one test, clean up others.

[WFLY-12298] Allow expressions; remove unnecessary code.

[WFLY-12298] Update WildFly Admin Guide.

[WFLY-12298] Various improvements

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [ ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue(s)
- [ ] Pull Request contains description of the issue(s)
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

ronsigal commented:
This pull request replaces https://github.com/wildfly/wildfly/pull/12543.

Hi @kabir , @bstansberry , @OndrejKotek 

I'm doing a pull request which, I hope, addresses most of the issues you raised in https://github.com/wildfly/wildfly/pull/12543. There's still a bit to do, but I want to put it out there now because I'm going to be off work for a few days.

It looks like these are the open issues:

1. Use of '-' instead of '.' in attribute names.

   Done
   
2. Assign attributes to an attribute group.

   Done
   
3. Define each attribute separately and explicitly.

   Done
   
4. Use ModelNode.FALSE / ModelNode.TRUE instead of Boolean.FALSE / Boolean.TRUE

   Done
   
5. Use 4 spaces for indentation

   Done
   
6. Create a service to hold attribute values instead of a static field.

   Done

7. Add attributes under a subresource (optional)

   Didn't do
   
Still to do:

1. transformers

2. test failures

kabir reviewed (commented):
I have had a quick look, but not done a full review

ronsigal commented on jaxrs/src/test/java/org/jboss/as/jaxrs/JaxrsSubsystem20TestCase.java:
```diff
@@ -0,0 +1,99 @@
+    protected String getSubsystemXml() throws IOException {
+        return
+              "<subsystem xmlns=\"urn:jboss:domain:jaxrs:2.0\">\n" +
+              "        <jaxrs-2-0-request-matching>true</jaxrs-2-0-request-matching>\n" +
```
Ok, thanks, @kabir.

ronsigal replied:
Done.

ronsigal commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/JAXRS.adoc:
```diff
@@ -0,0 +1,57 @@
+
+<subsystem xmlns="urn:jboss:domain:jaxrs:2.0">
+    <context-parameters>
+        <context-parameter name="resteasy.add.charset" value="true"/>
```
Ah, right. I just added documentation to my TODO list.

ronsigal replied:
Documentation updated.

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,237 @@
+ */
+public interface  JaxrsAttribute {
+
+    public static final String RESTEASY_PARAMETER_GROUP = "resteasy-parameter-group";
```
Well, currently, RESTEasy 3.x gets configured by context-parameters, but I've proposed incorporating MicroProfile Config into 3.x, so that there would be multiple configuration sources. That's why I used the more generic "resteasy-parameter-group".

jmesnil replied:
I'm not sure to understand what you mean.
This attribute group is a hint to group related attributes together. If RestEasy would be configured with MicroProfile Config, what would be the impact on these attributes? Would there be another set of attributes with the "mp-config" attribute group?
As an aside, the attribute group name could simply be "resteasy"

ronsigal replied:
I would imagine that there would be one group consisting of exactly the RESTEasy parameters. I was just reacting to the **name** "context-parameters" since it would be too specific once RESTEasy starts using MP Config (although it turns out that that won't happen in WF 19). Just using "resteasy" sounds fine to me.

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsContextParamHandler.java:
```diff
@@ -0,0 +1,271 @@
+   }
+
+   public static void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+      for (Entry<String, Boolean> entry : contextParametersBoolean.entrySet()) {
```
Aaarrgghh. That file has gone away, but the deletion didn't show up in my github branch. Sorry for wasting your time.

ronsigal replied:
JaxrsContextParamHandler deleted.

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsContextParamHandler.java:
```diff
@@ -0,0 +1,271 @@
+
+       @Override
+       public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+           context.getResult().set(JaxrsContextParamHandler.getContextParam(operation.get("name").asString()));
```
Sorry, again. JaxrsContextParamHandler is replaced by 

```
final class JaxrsParamHandler extends AbstractWriteAttributeHandler<Void> {
```
and JaxrsExtension registers it here:
```
    private static void registerAttributes(ManagementResourceRegistration registration) {
       for (SimpleAttributeDefinition definition : JaxrsAttribute.Attributes) {
          registration.registerReadWriteAttribute(definition, null, new JaxrsParamHandler(definition));
       }
    }
```

ronsigal commented:
Galleon Linux failed with "Failed to execute goal org.wildfly.build:wildfly-server-provisioning-maven-plugin:1.2.12.Final:build (server-provisioning) on project wildfly-build-legacy: Execution server-provisioning of goal org.wildfly.build:wildfly-server-provisioning-maven-plugin:1.2.12.Final:build failed: java.lang.RuntimeException: java.lang.IllegalStateException: Could not resolve '[subsystem=jaxrs.xml, supplement=null] Check the module for the extension has been defined." 

Do I have to do something?

kabir commented:
@ronsigal Regarding the failing build I think the problem is your splitting of subsystem-templates/jaxrs_1_0.xml and jaxrs_2_0.xml. Get rid of the 1.0 version, and rename the 2.0 one to simply jaxrs.xml. We just do the latest version, and these subsystem templates are used by the legacy feature-pack build (in the feature-pack/ folder). This is used by build-legacy and dist-legacy, which in turn are still used for some of the layered products.

It is good to put those values in there, but you also need to modify the galleon-pack/ stuff to set the same values for the subsystem. Otherwise (yes, we have some duplication at the moment to support both approaches) the built server that you end up with in build/ and dist/ will be different from the server in the xxx-legacy/ folders. The first set use galleon, while the second use the legacy feature packs.

Looking around the galleon-pack sources, I find the following two occurrances:
```
$git grep subsystem.jaxrs -- galleon-pack/
galleon-pack/src/main/resources/feature_groups/basic-profile.xml:    <feature spec="subsystem.jaxrs"/>
galleon-pack/src/main/resources/layers/standalone/jaxrs/layer-spec.xml:    <feature spec="subsystem.jaxrs"/>
```
These simply define the subsystem inline in the relevant layers/groups, which was fine since the subsystem was empty. However, now that you have some configuration I think you need to put   
```
<feature spec="subsystem.jaxrs"/>
```
into its own `<feature-group-spec` in its own file in the feature-groups/ folder in galleon-pack/ so that you configure it once (using `<param.../>` for the various attributes), and then reference it from basic-profile.xml and jaxrs/layer-spec.xml. To reference rather than declare, I believe the syntax is `<feature-group name="jaxrs"/>` (assuming your new feature-group-spec, containing the `subsystem.jaxrs` feature spec is called 'jaxrs').

ronsigal commented:
Thanks, @kabir. Right now I feel like a chimpanzee with marginally above average pattern matching skills. ;-)

1. So I can delete JaxrsSubsystemTestCase, right? (Which I've done.)

2. I've created galleon-pack/src/main/resources/feature_groups/jaxrs.xml and the parameters are showing up in build/target/wildfly-18.0.0.Beta1-SNAPSHOT/standalone/configuration/standalone.xml, etc. But I don't see in the xxx-legacy/ folders where to look for comparison ... .

3. Also, I don't know what the philosophy for <subsystem xmlns="urn:jboss:domain:jaxrs:2.0"> should be for these files. Right now I've got

```
        <subsystem xmlns="urn:jboss:domain:jaxrs:2.0">
            <jaxrs-2-0-request-matching>false</jaxrs-2-0-request-matching>
            <resteasy-add-charset>true</resteasy-add-charset>
            <resteasy-buffer-exception-entity>true</resteasy-buffer-exception-entity>
            <resteasy-disable-html-sanitizer>false</resteasy-disable-html-sanitizer>
            <resteasy-document-expand-entity-references>false</resteasy-document-expand-entity-references>
            <resteasy-document-secure-disableDTDs>true</resteasy-document-secure-disableDTDs>
            <resteasy-document-secure-processing-feature>true</resteasy-document-secure-processing-feature>
            <resteasy-gzip-max-input>10000000</resteasy-gzip-max-input>
            <resteasy-rfc7232preconditions>false</resteasy-rfc7232preconditions>
            <resteasy-role-based-security>false</resteasy-role-based-security>
            <resteasy-secure-random-max-use>100</resteasy-secure-random-max-use>
            <resteasy-use-builtin-providers>true</resteasy-use-builtin-providers>
            <resteasy-use-container-form-params>false</resteasy-use-container-form-params>
            <resteasy-wider-request-matching>false</resteasy-wider-request-matching>
        </subsystem>
```
That is, I've got exactly all of the parameters that have a default value. So, they're not necessary except as illustration. I'm not sure how to think about it.

-Ron

ronsigal commented:
I'm trying to re-run the tests manually, but it looks like they're running with an old version of wildfly-jaxrs ... . I'm going to create a dummy commit to see if the tests will start automatically.

ronsigal commented:
I've got a test that passes for me at home but fails here. I'm passing in the JNDI name of an EJB ("java:app/jaxrsnoap/EJB_Resource1"), and it's not getting found. At home, I can see 
```
2019-12-19 19:33:03,071 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-8) WFLYEJB0473: JNDI bindings for session bean named 'EJB_Resource1' in deployment unit 'deployment "jaxrsnoap.war"' are as follows:

	java:global/jaxrsnoap/EJB_Resource1!org.jboss.as.test.integration.jaxrs.cfg.EJB_Resource1
	java:app/jaxrsnoap/EJB_Resource1!org.jboss.as.test.integration.jaxrs.cfg.EJB_Resource1
	java:module/EJB_Resource1!org.jboss.as.test.integration.jaxrs.cfg.EJB_Resource1
	ejb:/jaxrsnoap/EJB_Resource1!org.jboss.as.test.integration.jaxrs.cfg.EJB_Resource1
	java:global/jaxrsnoap/EJB_Resource1
	java:app/jaxrsnoap/EJB_Resource1
	java:module/EJB_Resource1
```
in server.log, but here I get the error
```
Caused by: javax.naming.NameNotFoundException: jaxrsnoap/EJB_Resource1
```
Is there a way to see server.log here?

ronsigal commented:
I've done some experiments, and it seems that EJB_Resource1 and EJB_Resource2 aren't registered in JNDI. Maybe I need to do it manually ... .

ronsigal commented:
This is taking too much time. For now, I'm commenting out the "resteasy.jndi.resources" test.

ronsigal commented:
New problem. Some opentracing test cases are failing. They don't seem to have anything to do with RESTEasy or wildfly-jaxrs, but they fail only after the changes I've made to wildfly-jaxrs.

ronsigal commented:
I solved the problem with the opentracing tests, which leaves one problem: the failure of the EJB/JNDI test. I created a new pull request (https://github.com/wildfly/wildfly/pull/12866), which is the same as this one but based on the current (at the time) state of the master branch, just in case I was missing something. That didn't help. When I comment out that test, everything passes. I've posted a discussion of the issue on https://issues.redhat.com/browse/EAP7-634 and mentioned the problem to Ondrej Kotek in particular.

kabir commented on jaxrs/src/test/java/org/jboss/as/jaxrs/JaxrsSubsystemTestCase.java:
```diff
@@ -1,62 +0,0 @@
-/**
- * @author <a href="kabir.khan@jboss.com">Kabir Khan</a>
- */
-public class JaxrsSubsystemTestCase extends AbstractSubsystemBaseTest {
```
This test should not be deleted. We need to make sure that the old parsers work for backward compatibility. But rename it to include the model version 1.0

ronsigal replied:
Ok, I've restored JaxrsSubsystemTestCase, @kabir. Is it legitimate to override testSubsystem() as
```
    @Test
    @Override
    public void testSubsystem() throws Exception {
        standardSubsystemTest(null, false);  // <<== don't compare XML
    }
```
to avoid comparing XML, since otherwise it fails with
```
org.junit.ComparisonFailure: expected:<...:jboss:domain:jaxrs:[1].0"/>
> but was:<...:jboss:domain:jaxrs:[2].0"/>
>
```
I see similar usage in legacy tests in other subsystem.s

kabir commented on jaxrs/src/test/java/org/jboss/as/jaxrs/JaxrsSubsystem20TestCase.java:
```diff
@@ -0,0 +1,161 @@
+
+    @Test
+    public void testTransformers() throws Exception {
+        testTransformers_2_0_to_1_0(ModelTestControllerVersion.EAP_7_2_0);
```
We generally (although it doesn't happen *everywhere*!) use two tests for these, e.g:

* testTransformers720()
* testRejectingTransformers720()

Also, it is typically better to pull these out into a new JaxrsTransformersTestCase class since we need parser test for each model version (see my comment about deleting JaxrsSubsystemTestCase.java)

ronsigal replied:
@kabir, I've broken out JaxrsTransformersTestCase. When you talk about testTransformers720() and testRejectingTransformers720(), are you just talking about renaming the tests? That's what I've done, in fact.

kabir reviewed (changes requested):
I still haven't reviewed it in-depth, I just looked at a few things in one of your comments which made me react :)

kabir commented on docs/src/main/asciidoc/_extending-wildfly/Domain_Mode_Subsystem_Transformers.adoc:
```diff
@@ -271,8 +271,8 @@ Once it has been increased you can of course make more changes until the
-For example the `jaxrs` subsystem has remained on 1.0.0 for all versions
-of WildFly and JBoss AS 7.
+For example the `jaxrs` subsystem remained on 1.0.0 for all versions
+of WildFly and JBoss AS 7 through Wildfly 18.
```
Good catch!

darranl commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,202 @@
+package org.jboss.as.jaxrs;
```
Missing copyright header

ronsigal replied:
Done. Thanks, @darranl.

darranl commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsConstants.java:
```diff
@@ -0,0 +1,56 @@
+package org.jboss.as.jaxrs;
```
Missing copyright header

ronsigal replied:
Done.

darranl commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,59 @@
+            version="1.0">
+
+    <!-- The jaxrs subsystem root element -->
+    <xs:element name="subsystem" type="subsystemType"/>
```
If bumping the model version of a subsystem it is better to handle the change as two commits, otherwise it is not possible to compare the changes actually made to the new schema without manually diffing the files.

ronsigal replied:
I see. Got it. Thanks.

darranl reviewed (changes requested):
Don't worry about the double commit comment for the schema, after manually checking I see it was empty but worth keeping in mind for future PRs.

Have added a comment in a couple of new files missing a copyright header.

ronsigal commented:
Copying @kabir's comment from https://github.com/wildfly/wildfly/pull/12866:

@ronsigal I am getting a bit confused about the need for this new PR :) I think you can just rebase your original on upstream/master and force push?

Re the EJB test failures, it might due to the fact that testsuite/integration/basic/pom.xml has two main surefire executions:

    basic-integration-default-full.surefire - uses standalone-full.xml
    basic-integration-default-web.surefire - uses standalone.xml

My bet is that your ResteasyAttributeTestCase gets run in the standalone.xml one, which AFAICR does not have EJB, so perhaps that is it.

Next I see a 'layers.profile' in the same pom, which is activated with -Dts.layers. It seems to have 'cloud-profile-layers.surefire' execution which tests against a Galleon slimmed server. This already excludes a lot of tests relying on EJB, so if the above theory is correct you should exclude this test from there too.

I will close this PR so we can continue discussion on the other one. If it is not possible to get the other PR in the same shape as this one, feel free to reopen this one (or ask me if you don't have permissions).

ronsigal commented:
I have excluded ResteasyAttributeTest from the basic-integration-default-web.surefire execution, and now ResteasyAttributeTestCase is failing only in the "Galleon Linux - JDK 11" build. @kabir, maybe that has something to do with your other comment, but I'm not seeing a "layers.profile" anywhere ...

ronsigal commented:
Now there are a bunch of StandardConfigsXMLValidationUnitTestCase failures. I haven't figured that out yet.

kabir commented:
@ronsigal `git grep layers.profile` takes me to https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/pom.xml#L805

The 'Galleon Linux - JDK 11' job uses `-Dts.layers` so I think your theory is correct :)

kabir commented:
```
org.wildfly.dist.legacy.subsystem.xml.StandardConfigsXMLValidationUnitTestCase
```
is in dist-legacy

kabir commented on jaxrs/src/main/resources/subsystem-templates/jaxrs.xml:
```diff
@@ -3,4 +3,4 @@
 <config>
    <extension-module>org.jboss.as.jaxrs</extension-module>
    <subsystem xmlns="urn:jboss:domain:jaxrs:1.0"/>
-</config>
```
This should probably reference 2.0

ronsigal commented:
Thanks, @kabir. I just found it a few minutes ago and was able to reproduce the failures. I think I mis-followed your instructions about subsystem-templates. I'm checking on that now.

ronsigal commented:
re: "This should probably reference 2.0"

That's exactly what I was looking at!!

ronsigal commented:
Good news. StandardConfigsXMLValidationUnitTestCase just passed, so that's that.

ronsigal commented:
re: "The 'Galleon Linux - JDK 11' job uses -Dts.layers so I think your theory is correct :)"

Umm, did I have a theory?

ronsigal commented:
re: "@ronsigal git grep layers.profile takes me to https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/pom.xml#L805"

Ah, I was looking at an old version of that file.

ronsigal commented:
Hmmm. I put JaxrsSubsystemTestCase back in, but now testSubsystem() is failing because I changed subsystem-templates/jaxrs.xml. I'm not sure how to handle this.

ronsigal commented:
Well, this is good. I think all of the major issues are resolved, as far as I can tell. At least until you take a close look, @kabir. ;-)

1. One concern I have is, I don't really know what subsystem-templates/jaxrs.xml should look like. Right now, it has an entry for every parameter that has a default value, but it could just as well be empty. I don't know what the convention is for that.

2. I've noticed a couple of parameters I didn't see before. Maybe I was looking at a different version of RESTEasy. Anyway, I'll fix that tomorrow.

-Ron

kabir commented:
@ronsigal I still haven't looked but just to answer your question. The legacy and the new build should result in the same xml.

If you just want to leave the parameters empty in the configuration, that is fine. However, in the subsystem test you should aim for full coverage (i.e. add all the parameters to the xml). In that case it might make sense to do an additional minimal subsystem test as well (i.e. empty xml), and then check that the results of a /subsystem=jaxrs:read-resource(include-defaults=true) gives the expected values

ronsigal commented:
Thanks, @kabir. I have 

1. Minimized the jaxrs subsystem config and I get the same minimal subsystem config in dist, dist-legacy, and build.

2. Added a JaxrsSubsystem20MinimalTestCase

3. Checked that /subsystem=jaxrs:read-resource(include-defaults=true) looks good.

-Ron

ronsigal commented:
Recent changes:

1. After excluding ResteasyAttributeTest from the basic-integration-default-web.surefire execution, I realized that the jaxrs tests weren't included in the basic-integration-default-full.surefire execution, so I fixed that.

2. I added the "resteasy.preferJacksonOverJsonB" parameter.

ronsigal commented:
Hi @kabir,

A little history.

1. My first attempt, with a myriad of associated problems, is in pull request https://github.com/wildfly/wildfly/pull/12543. I started over in this current pull request, beginning with a list (https://github.com/wildfly/wildfly/pull/12835#issuecomment-561932418) of all of the things I had done in response.

2. For the current pull request, the only substantive problems you highlighted were due to my accidental inclusion of a discarded class. Other than that, I have contended with a variety of test failures, and you have stepped in each time with invaluable suggestions.

As of now, I have addressed every problem that I am aware of, and I am anxiously awaiting the next round of issues. ;-) I'm hoping there's nothing serious. I can't do a RESTEasy 3.10.0.Final release for WF 19 until I know that this issue is approved (or not), and I'd like to submit an upgrade pull request as soon as possible.

-Ron

kabir reviewed (changes requested):
Hi @ronsigal I have done an initial  pass, in this round I was mainly doing a fine-comb comparison of the xsd, the attributes and the stuff in the https://docs.jboss.org/resteasy/docs/3.8.1.Final/userguide/html/Installation_Configuration.html#configuration_switches document.

Mainly you are missing a few defaults, which I have commented in the xsd.

Next there are some attributes which are in a String in the model, but which are actually lists or maps. Generally we implement those in the model with SimpleListAttributeDefinition.Builder and something like PropertyObjectTypeAttributeDefinition.Builder. This allows us to add/remove from these attributes without overwriting the whole value. See https://github.com/wildfly/wildfly-core/blob/master/controller/src/test/java/org/jboss/as/controller/operation/global/CollectionOperationsTestCase.java for an example of what I am talking about :) You will most likely need to change your xsd for these attributes so that the xml looks something like
**List**
`<my-list-element>first second third</my-list-element>` (if xs:list is valid for xs:element. For xs:attribute it definitely is, and xml lists in attributes are space separated)
or perhaps
```
<my-list-element>
   <value>one</value>
   <value>two</value>
</my-list-element>
```

For Maps, I think something like
```
<my-map-element>
   <entry key="a" value="one"/>
   <entry key="b" value="two"/>
</my-map-element>
```
or perhaps
```
<my-map-element>
   <entry key="a">one</entry>
   <entry key="b">two</entry>
</my-map-element>
```
Look for usages of those builders I used to give you some inspiration for how we normally handle this (I don't remember the **exact** details off the top of my head).

Finally, I noticed that we have some attributes which do not appear in the https://docs.jboss.org/resteasy/docs/3.8.1.Final/userguide/html/Installation_Configuration.html#configuration_switches document. This may be fine (and I have hazy memories about having spoken about them ages ago :) ), but since I am here, they are:

* RESTEASY_DISABLE_PROVIDERS
* RESTEASY_PREFERJACKSONOVERJSONB

Then conversely we have some which appear in that document, but not in the model. They are (and again it might be fine):

* resteasy.servlet.mapping.prefix
* resteasy.scan
* resteasy.scan.providers
* resteasy.scan.resources
* javax.ws.rs.Application

ronsigal commented:
re: " we have some attributes which do not appear ... "

I did an inventory of all of the existing parameters yesterday. 

* RESTEASY_PREFERJACKSONOVERJSONB exists ... it's described in the JSONB chapter
* RESTEASY_DISABLE_PROVIDERS is new for this issue - it was proposed in the initial description of EAP7-634
* resteasy.scan, resteasy.scan.providers, resteasy.scan.resources are no longer in service, and I've removed them from the table
* For now, we're just doing parameters that apply to all deployments, so resteasy.servlet.mapping.prefix and javax.ws.rs.Application don't seem relevant.

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsConstants.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String RESTEASY_LANGUAGE_MAPPINGS = "resteasy-language-mappings";
+    public static final String RESTEASY_MEDIA_TYPE_MAPPINGS = "resteasy-media-type-mappings";
+    public static final String RESTEASY_MEDIA_TYPE_PARAM_MAPPING = "resteasy-media-type-param-mapping";
+    public static final String RESTEASY_PREFERJACKSONOVERJSONB = "resteasy-preferJacksonOverJsonB";
```
The name of this sticks out a bit - is it possible to make it follow the normal conventions? e.g. resteasy-prefer-jackson-jsonb

ronsigal replied:
Yeah, I agree. I can change RESTEASY_PREFERJACKSONOVERJSONB in wildfly-jaxrs, but, unfortunately, "resteasy.preferJacksonOverJsonB" is already defined in RESTEasy.

Done.

kabir commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-language-mappings" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-media-type-mappings" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-media-type-param-mapping" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-preferJacksonOverJsonB" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Name sticks out a bit, as mentioned in JaxrsConstants

jmesnil commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAllowExpression(true)
+            .setValidator(new ModelTypeValidator(ModelType.BOOLEAN, false))
+            .setDefaultValue(ModelNode.FALSE)
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
```
Given the prefix of this attributes (`jaxrs-2`), does it really belong to the RESTEASY_PARAMETER_GROUP attribute group (which to me signals attributes that are specific to the RestEasy implementation)?

ronsigal replied:
Hi @jmesnil,

Sigh. I was about to say, it's a bad name, I have no idea where that came from. Then I did a little search and discovered that I named it myself for https://issues.redhat.com/browse/RESTEASY-1951. It's meant to provide backwards compatibility with the JAX-RS 2.0 request matching algorithm. So, it's about JAX-RS but it's specific to RESTEasy. I should have named it more conventionally.

-Ron

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsServerConfig.java:
```diff
@@ -0,0 +1,186 @@
+ */
+public class JaxrsServerConfig {
+
+    private ModelNode jaxrs20RequestMatching;
```
I don't think we should use ModelNode for the fields in this class. The simple types should be enough?

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsValidator.java:
```diff
@@ -0,0 +1,39 @@
+import org.jboss.as.jaxrs.logging.JaxrsLogger;
+import org.jboss.dmr.ModelNode;
+
+public class JaxrsValidator implements ParameterValidator {
```
I guess this class will not be needed once you do the LIST/OBJECT changes?

ronsigal replied:
Gone.

kabir commented on jaxrs/src/test/java/org/jboss/as/jaxrs/JaxrsTransformersTestCase.java:
```diff
@@ -0,0 +1,152 @@
+
+    @Test
+    public void testRejectingTransformers720() throws Exception {
+        testRejections_2_0_0(ModelTestControllerVersion.EAP_7_2_0);
```
Nitpick: I'd call this testRejections_1_0_0 (since that is the ModelVersion that is doing the rejecting)

ronsigal replied:
Done.

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jaxrs/cfg/ResteasyAttributeTestCase.java:
```diff
@@ -0,0 +1,365 @@
+     * Reset attributes to default values.
+     */
+    static void resetAttributeValues() throws IOException {
+        setAttributeValue(JaxrsAttribute.JAXRS_2_0_REQUEST_MATCHING, JaxrsAttribute.JAXRS_2_0_REQUEST_MATCHING.getDefaultValue());
```
Not a huge deal, but instead of setting to the defaults you could just call `/subsystem=jaxrs:undefine-attribute(name=blah)` for each of these.

ronsigal replied:
Done

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jaxrs/cfg/ResteasyAttributeTestCase.java:
```diff
@@ -0,0 +1,365 @@
+        setAttributeValue(JaxrsAttribute.RESTEASY_DOCUMENT_SECURE_DISABLE_DTDS, JaxrsAttribute.RESTEASY_DOCUMENT_SECURE_DISABLE_DTDS.getDefaultValue());
+        setAttributeValue(JaxrsAttribute.RESTEASY_DOCUMENT_SECURE_PROCESSING_FEATURE, JaxrsAttribute.RESTEASY_DOCUMENT_SECURE_PROCESSING_FEATURE.getDefaultValue());
+        setAttributeValue(JaxrsAttribute.RESTEASY_GZIP_MAX_INPUT, JaxrsAttribute.RESTEASY_GZIP_MAX_INPUT.getDefaultValue());
+        setAttributeValue(JaxrsAttribute.RESTEASY_JNDI_RESOURCES, new ModelNode(""));
```
Actually this is wrong and should be undefine-attribute

ronsigal replied:
Done

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/management/api/SimpleSubsystemsTestCase.java:
```diff
@@ -43,6 +44,10 @@
     @Test
+    @Ignore
+    /*
+     * The jaxrs subsystem is no long empty.
```
'no long**er**' - important I know :)

ronsigal replied:
Oh noooo!!! I might have to push that off to WF 20!!

ronsigal replied:
Ok, I went the extra mile. ;-)

kabir reviewed (changes requested):
Hi Ron,

I think I am done with this round of review now :)

@asoldano Can you confirm that the tests in this PR cover what is mentioned in the test plan?

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+    <xs:element name="subsystem" type="subsystemType"/>
+    <xs:complexType name="subsystemType">
+       <xs:all minOccurs="0" maxOccurs="1">
+          <xs:element name="jaxrs-2-0-request-matching" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+    <xs:complexType name="subsystemType">
+       <xs:all minOccurs="0" maxOccurs="1">
+          <xs:element name="jaxrs-2-0-request-matching" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-add-charset" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+       <xs:all minOccurs="0" maxOccurs="1">
+          <xs:element name="jaxrs-2-0-request-matching" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-add-charset" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-buffer-exception-entity" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="jaxrs-2-0-request-matching" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-add-charset" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-buffer-exception-entity" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-disable-html-sanitizer" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-buffer-exception-entity" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-disable-html-sanitizer" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-disable-providers" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-expand-entity-references" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-disable-html-sanitizer" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-disable-providers" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-expand-entity-references" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-secure-disableDTDs" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-disable-providers" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-expand-entity-references" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-secure-disableDTDs" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-secure-processing-feature" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-document-expand-entity-references" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-secure-disableDTDs" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-document-secure-processing-feature" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-gzip-max-input" type="xs:integer" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-preferJacksonOverJsonB" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-providers" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-resources" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-rfc7232preconditions" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-providers" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-resources" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-rfc7232preconditions" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-role-based-security" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-resources" type="xs:string" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-rfc7232preconditions" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-role-based-security" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-secure-random-max-use" type="xs:integer" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-rfc7232preconditions" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-role-based-security" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-secure-random-max-use" type="xs:integer" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-use-builtin-providers" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/resources/schema/jboss-as-jaxrs_2_0.xsd:
```diff
@@ -0,0 +1,60 @@
+          <xs:element name="resteasy-role-based-security" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-secure-random-max-use" type="xs:integer" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-use-builtin-providers" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
+          <xs:element name="resteasy-use-container-form-params" type="xs:boolean" minOccurs="0" maxOccurs="1"/>
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_DISABLE_PROVIDERS = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_DISABLE_PROVIDERS, ModelType.STRING)
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_JNDI_RESOURCES = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_JNDI_RESOURCES, ModelType.STRING)
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_LANGUAGE_MAPPINGS = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_LANGUAGE_MAPPINGS, ModelType.STRING)
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_MEDIA_TYPE_MAPPINGS = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_MEDIA_TYPE_MAPPINGS, ModelType.STRING)
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_PROVIDERS = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_PROVIDERS, ModelType.STRING)
```
Done

ronsigal commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsAttribute.java:
```diff
@@ -0,0 +1,232 @@
+            .setAttributeGroup(RESTEASY_PARAMETER_GROUP)
+            .build();
+
+    SimpleAttributeDefinition RESTEASY_RESOURCES = new SimpleAttributeDefinitionBuilder(JaxrsConstants.RESTEASY_RESOURCES, ModelType.STRING)
```
Done

ronsigal commented:
I've committed a number of changes suggested by @kabir, largely related to re-implementing list and set attributes properly.

kabir commented:
@ronsigal I have taken a quick look (on my phone in a hostelry :) ) and your first commit looks on the right track. I guess the next step is to convert the lists and objects into something better for the holder class for the config service?

The second merge commit I don’t really get, I guess that is trying to update your branch with the latest from master. Which may or may not be fine. @darranl is the expert at that, I normally rebase. The reason I mention this is really the test failures that I have not looked at. They may well be related to what I say in the previous paragraph.

ronsigal commented:
Good morning, @kabir. After I submitted my commit yesterday, I realized that I hadn't updated the integration test ResteasyAttributeTestCase, which has compilation errors now. I'm still working on updating it - had gremlins last night.

As for the merger from master, I don't know how that happened. It's not there on my laptop. Sometimes I get a conflict in testsuite/integration/basic/pom.xml which I just fix on github. Could that have done something weird? I don't know ...

kabir commented:
@ronsigal I saw that merge commit here, got rid of it and pushed a rebased version to my help-ron branch. There were a few conflicts (in testsuite/integration/basic/pom.xml). Fetch my branch and make sure it looks ok, and then `git reset --hard kabir/help-ron`

ronsigal commented:
Thanks, @kabir. I'll look at that tonight or tomorrow. For now, I've fixed ResteasyAttributeTestCase and removed "resteasy.resources". Still have a bit more to clean up, but most things are done.

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jaxrs/cfg/ResteasyAttributeTestCase.java:
```diff
@@ -153,24 +155,41 @@ public static void afterClass() throws IOException {
         setAttributeValue(JaxrsAttribute.RESTEASY_DISABLE_HTML_SANITIZER, changeBoolean(JaxrsAttribute.RESTEASY_DISABLE_HTML_SANITIZER.getDefaultValue()));
-        setAttributeValue(JaxrsAttribute.RESTEASY_DISABLE_PROVIDERS, new ModelNode(Provider1.class.getName() + "," + Provider2.class.getName()));
+        ModelNode list = new ModelNode().setEmptyList();
+        list.add(FileProvider.class.getName());
```
Here and in other attributes that support expressions, you should do something like:
```
ModelNode value = new ModelNode(new ValueExpression("${rest.test.dummy.property:" + FileProvider.class.getName() + "}"));
list.add(value);
```
This ValueExpression stuff seems quite new, the old way which is deprecated but still heavily used is
```
ModelNode value = new ModelNode();
value.setExpression("${rest.test.dummy.property:" + FileProvider.class.getName() + "}")
```

ronsigal replied:
Done. I put expressions in one list and one map.

kabir replied:
I think you should use expressions everywhere where they are supported. It helps 

- make sure that they actually work in the model (although the subsystem test is good enough for that)
- make sure that they are translated/resolved properly for the runtime (not really possible in the subsystem tests)

ronsigal replied:
Ok

ronsigal replied:
Done

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jaxrs/cfg/ResteasyAttributeTestCase.java:
```diff
@@ -207,11 +226,32 @@ static void resetAttributeValues() throws IOException {
+                break;
+
+            default:
+                expectedValue = value;
```
I believe here that you are checking what RestEasy pulls out of itself rather than the model values. The former will have resolved expressions, while the latter will have the full expression.

So in order to not compare apples and oranges, I think here you need to do something like
```
expectedValue = value.resolve();
```
This does more or less the same as the AttributeDefinition.resolve() stuff earlier, but does not have the extra stuff. Still as this is test code it should be enough.

You will need to do the same in your textifyList/Map() methods

ronsigal replied:
Thanks, @kabir. I didn't know I had to deal explicitly with expressions. I'll look into that.

ronsigal replied:
Done

kabir reviewed (changes requested):
@ronsigal It looks good so far, but I just realised something else when reviewing. You are not handling expressions in your code as I've tried to explain in my comments.

ronsigal commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/JAXRS.adoc:
```diff
@@ -0,0 +1,58 @@
+    <resteasy-media-type-param-mapping>mt</resteasy-media-type-param-mapping>
+    <resteasy-providers>com.bluemonkey.reader,com.bluemonkey.writer</resteasy-providers>
+</subsystem>
+
```
Done (in the forthcoming commit).

ronsigal commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/JAXRS.adoc:
```diff
@@ -0,0 +1,58 @@
+5. map
+
+The syntax of the first three is as usual. A list is comma separated, and
+a map is a comma separated list of "key:value" elements. White space is ignored.
```
This text is now re-written.

ronsigal commented:
Hi @kabir,

Regarding expressions, I see calls to JaxrsServerConfig.setXXX() methods in two places:

1. JaxrsSubsystemAdd.createServerConfig(), where the calls look like
```
if (configuration.hasDefined(JaxrsConstants.RESTEASY_DISABLE_PROVIDERS)) {
   config.setResteasyDisableProviders(JaxrsAttribute.RESTEASY_DISABLE_PROVIDERS.resolveModelAttribute(context, configuration));
}
```
and
2. JaxrsParamHandler.updateServerConfig(), which is called from two places:

2a. 
```
protected boolean applyUpdateToRuntime(OperationContext context, ModelNode operation, String attributeName,
        ModelNode resolvedValue, ModelNode currentValue, HandbackHolder<Void> handbackHolder)
                throws OperationFailedException {

    //If the required value is the current one, we do not need to do anything.
    if (!isSameValue(context, resolvedValue, currentValue, attributeName)) {
        final String value = resolvedValue.isDefined() ? resolvedValue.asString() : null;
        updateServerConfig(context, attributeName, value, false);
    }
    return false;
}
```
and 

2.b. 
```
protected void revertUpdateToRuntime(OperationContext context, ModelNode operation, String attributeName,
        ModelNode valueToRestore, ModelNode valueToRevert, Void handback) throws OperationFailedException {
    final String value = valueToRestore.isDefined() ? valueToRestore.asString() : null;
    updateServerConfig(null, attributeName, value, true);
}
```
According to the javadoc, 
```

ronsigal commented:
Regarding the types stored in JaxrsServerConfig, how strongly do you feel about using the specific types of each attribute? Ultimately, they end up with JaxrsIntegrationProcessor passing Strings into RESTEasy, so being able to call modelNode.asString() in JaxrsIntegrationProcessor is kinda convenient ....

ronsigal commented:
@kabir, I've added some expression testing to ResteasyAttributeTestCase. It works without any changes to wildfly-jaxrs, so maybe my conjecture is correct?

Wow, it took me a while to figure out how to set the system properties early enough. Turns out this works:
```
    static void setServerSystemProperty(String property, String value) throws IOException {
        ModelNode address = Operations.createAddress("system-property", property);
        ModelNode op = Operations.createAddOperation(address);
        op.get(VALUE).set(value);
        ModelNode result = modelControllerClient.execute(op);
        Assert.assertEquals(SUCCESS, result.get(OUTCOME).asString());
    }
```

kabir commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/JAXRS.adoc:
```diff
@@ -0,0 +1,113 @@
+*Note.* Since there parameters are global, the classes referred in "resteasy.providers" and "resteasy.disable.providers"
+must be available to all deployments. In practice, then, they are meant to enable
+or disable RESTEasy providers. Note that they can be used in conjunction with
+"resteasy-use-builtin-providers" to tailor a set of available providers. 
```
Effectively (if I am understanding properly) there are only two places a provider can happen currently:

- In the deployment, which we discourage since those are not available to all deployments
- "Built-in" - I don't know how resteasy-use-builtin-providers relates to that.

(Additionally they could be specified via a drop-in module, but for that we would need to be able to specify the module the class is to be loaded from)

Since I don't know what resteasy-use-builtin-providers I don't know if it makes sense to keep resteasy.providers or not.

I realise this comment is a bit muddled but hopefully it triggers something for you :)

ronsigal replied:
Hi @kabir. RESTEasy comes with some enabled built-in providers, and they can all be disabled by setting "resteasy-use-builtin-providers" to "false". There is a discussion on https://issues.redhat.com/browse/EAP7-634 about that. The point was that disabling one provider meant disabling all of them and then adding the rest back in using "resteasy.providers". That was the motivation for introducing "resteasy.disable.providers".

ronsigal replied:
Btw, I think adding a "module" key would be great, but for next time.

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java:
```diff
@@ -458,5 +469,145 @@ public static void setContextParameter(JBossWebMetaData webdata, String name, St
+    }
+
+    private String convertListToString(ModelNode modelNode) {
+        StringBuffer sb = new StringBuffer();
```
StringBuffer is a bit 90s (all its methods are synchronized), use StringBuilder instead.

ronsigal replied:
Ok, I always forget what the new one's called.

ronsigal replied:
re; "final round"

In the words of Roberto Duran: "No mas."  Lol.

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java:
```diff
@@ -458,5 +469,145 @@ public static void setContextParameter(JBossWebMetaData webdata, String name, St
+    }
+
+    private String convertMapToString(ModelNode modelNode) {
+        StringBuffer sb = new StringBuffer();
```
s/StringBuffer/StringBuilder

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java:
```diff
@@ -458,5 +469,145 @@ public static void setContextParameter(JBossWebMetaData webdata, String name, St
+
+    private String convertListToString(ModelNode modelNode) {
+        StringBuffer sb = new StringBuffer();
+        String[] ss = strip(modelNode.asString().trim(), '[', ']').split(",");
```
Rather than do this, it is better to do
```
for (ModelNode value : modelNode.asList()) {
    // TODO Add comma depending on value of first

    sb.append(value.asString());
}

ronsigal replied:
Ah, ok.

ronsigal replied:
I realize that I had some old code in JaxrsParamHandler that was saving resolved ModelNodes as strings, so JaxrsIntegrationProcessor was getting strings instead of lists and objects. I fixed that, and now your simpler code works.

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java:
```diff
@@ -458,5 +469,145 @@ public static void setContextParameter(JBossWebMetaData webdata, String name, St
+
+    private String convertMapToString(ModelNode modelNode) {
+        StringBuffer sb = new StringBuffer();
+        String[] ss = strip(modelNode.asString().trim(), '{', '}').split(",");
```
Rather than this, it is better to do

```
        for (String key : modelNode.keys()) {
            ModelNode value = modelNode.get(key);

            // TODO first/comma stuff

            // I am not sure what the exact format should be, but for illustration this should give you an idea 
            sb.append(key + "=" + value.asString());
        }
```

ronsigal replied:
ok.

ronsigal replied:
Done.

kabir commented on jaxrs/src/main/java/org/jboss/as/jaxrs/JaxrsConstants.java:
```diff
@@ -0,0 +1,77 @@
+    public static final String RESTEASY_LANGUAGE_MAPPINGS = "resteasy-language-mappings";
+    public static final String RESTEASY_MEDIA_TYPE_MAPPINGS = "resteasy-media-type-mappings";
+    public static final String RESTEASY_MEDIA_TYPE_PARAM_MAPPING = "resteasy-media-type-param-mapping";
+    public static final String RESTEASY_PREFER_JACKSON_OVER_JSONB = "resteasy-preferJacksonOverJsonB";
```
I still hate the name of the content of this string :) E.g the following looks ugly:

```
/subsystem=jaxrs:read-attribute(name=resteasy-preferJacksonOverJsonB)
```
I think this should be changed so that it is called 'prefer-jackson-over-jsonb' in the model at least, and add some special handling for the translation to what RestEasy expects for this case.
For the xml it **might** be ok to keep the name like this if you have some concerns about it looking similar with RestEasy's value but I think it is better to be consistent in both places our end.

@jmesnil Do you agree?

ronsigal replied:
Ok, no problem. I can finesse it in JaxrsIntegrationProcessor.

ronsigal replied:
Done.

kabir commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jaxrs/cfg/ResteasyAttributeTestCase.java:
```diff
@@ -0,0 +1,475 @@
+    private static final String FAILED = "failed";
+    private static final String RESTEASY_MEDIA_TYPE_PARAM_MAPPING_CONTEXT_VALUE = "resteasy.media.type.param.mapping.context.value";
+
+    private static final ModelNode VALUE_EXPRESSION_BOOLEAN_TRUE = new ModelNode(new ValueExpression("${rest.test.true:fail}"));
```
This is ok, but your test will be a lot simpler if you just use the following string for the ValueExpression:
```
"${org.rest.test.dummy:true}"
```
Then you don't have to worry about setting all the system properties, instead it will fall back to the default value specified in the expression. The expression resolution mechanism is well tested elsewhere, so there is no need to repeat that here. The main thing is to make sure that the runtime stuff gets the proper values.

ronsigal replied:
Nice. Ok, thanks.

kabir reviewed (changes requested):
@ronsigal I think this is the final round :)

ronsigal commented:
@kabir, I've uploaded a commit with 

1. StringBuilder

2. Improved "stringify code"

3. Renamed "resteasy.preferJacksonOverJsonB"

I.e., everything but the simplified ValueExpressions, which I'm having a problem with just now ...

kabir commented:
Test failures are unrelated/intermittent

ronsigal commented:
Hmmm, in these latest failures, the line numbers indicate an earlier version. I fixed the problem they're failing with.  Aarrgghh.

ronsigal commented:
Going out for dinner ...

kabir commented:
There were so many CI jobs running, probably one for each recent push. I stopped all apart from the latest, and they look good so far (one of them has a unrelated Vdx failure)

ronsigal commented:
Ah, great. I don't really know how to manage these jobs.

ronsigal commented:
@kabir, I just noticed that I'm getting
```
<resteasy-add-charset>${dummy:false}</resteasy-add-charset>
```
in standalone.xml. I'm writing it with
```
    public void writeContent(final XMLExtendedStreamWriter streamWriter, final SubsystemMarshallingContext context) throws XMLStreamException {
        context.startSubsystemElement(JaxrsExtension.NAMESPACE_2_0, false);
        ModelNode subsystem = context.getModelNode();
        for (AttributeDefinition attr : JaxrsAttribute.ATTRIBUTES) {
            attr.marshallAsElement(subsystem, true, streamWriter);
        }
        streamWriter.writeEndElement();
    }
```
and it seems marshallAsElement() doesn't resolve the expression ...

ronsigal commented:
Oh, well, I see
```
<wsdl-host>${jboss.bind.address:127.0.0.1}</wsdl-host>
```
in the web services subsystem, so ... maybe that's ok?

ronsigal commented:
Looks like another intermittent failure.

jamezp commented:
> @kabir, I just noticed that I'm getting
> 
> ```
> <resteasy-add-charset>${dummy:false}</resteasy-add-charset>
> ```
> 
> in standalone.xml. I'm writing it with
> 
> ```
>     public void writeContent(final XMLExtendedStreamWriter streamWriter, final SubsystemMarshallingContext context) throws XMLStreamException {
>         context.startSubsystemElement(JaxrsExtension.NAMESPACE_2_0, false);
>         ModelNode subsystem = context.getModelNode();
>         for (AttributeDefinition attr : JaxrsAttribute.ATTRIBUTES) {
>             attr.marshallAsElement(subsystem, true, streamWriter);
>         }
>         streamWriter.writeEndElement();
>     }
> ```
> 
> and it seems marshallAsElement() doesn't resolve the expression ...

@ronsigal Do you want it as an attribute? If so look at `marshallAsAttribute()`.

ronsigal commented:
Well, I just couldn't get that elegant treatment of ValueExpressions to work. I went out for pizza, came home, puttered about, and now they work, for some reason. Coincidentally, it was a pizza capriciosa.

ronsigal commented:
@jamezp, given that I now see unresolved expressions in standalone.xml, e.g.,
```
<wsdl-host>${jboss.bind.address:127.0.0.1}</wsdl-host>
```
I guess it's ok.

kabir commented:
@ronsigal The output of the xml is basically the same as the input. Expressions are only resolved for the runtime. So for the wsdl example, it allows you to use the same standalone.xml on a bunch of servers, and to override that for each server by e.g passing in `-D jboss.bind.address=$BIND_ADDR`, where `$BIND_ADDR` could be from a user's wrapper script, or set up on the machine, contaner or whatever

kabir commented:
@ronsigal The subsystem tests should be converted to use expressions. That can come in later for WF19, I opened https://issues.redhat.com/browse/WFLY-12960 to track that

ronsigal commented:
re: "it allows you to use the same standalone.xml on a bunch of servers"

Ah, I see.

ronsigal commented:
re: " I opened https://issues.redhat.com/browse/WFLY-12960 to track that"

Great. So, I have a little bit to do in RESTEasy for "resteasy.disable.providers", then I can do a release and we can wrap this up. !!

ronsigal committed:
[WFLY-12298] Implement RESTEasy context parameters in Wildfly
